crates/cargo-contract/src/main.rs

```diff
@@ -486,15 +486,41 @@ fn main() {
     tracing_subscriber::fmt::init();
 
     let Opts::Contract(args) = Opts::parse();
-    let suppress_err = match &args.cmd {
-        Command::Call(call_cmd) if call_cmd.is_json() => true,
-        Command::Instantiate(inst_cmd) if inst_cmd.is_json() => true,
+    let is_json = match &args.cmd {
+        Command::Call(call_cmd) => call_cmd.is_json(),
+        Command::Instantiate(inst_cmd) => inst_cmd.is_json(),
+        Command::Upload(upld_cmd) => upld_cmd.is_json(),
         _ => false,
     };
     match exec(args.cmd) {
         Ok(()) => {}
         Err(err) => {
-            if !suppress_err {
+            if is_json {
+                // error message can be either plain text or json string
+                // we need to check whether the error message is json object
+                let json_result =
+                    serde_json::from_str::<serde_json::Value>(&err.to_string());
```

Comment:
It's not great to serialize to json at the error site and then deserialize again at the top level. I think we can do better with a strongly typed approach see https://github.com/paritytech/cargo-contract/pull/736